src/main/java/org/spongepowered/common/data/processor/data/tileentity/ShulkerBoxDyeableDataProcessor.java

```diff
@@ -0,0 +1,114 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.data.processor.data.tileentity;
+
+
+import net.minecraft.item.EnumDyeColor;
+import net.minecraft.tileentity.TileEntityShulkerBox;
+import org.spongepowered.api.Sponge;
+import org.spongepowered.api.data.DataContainer;
+import org.spongepowered.api.data.DataHolder;
+import org.spongepowered.api.data.DataTransactionResult;
+import org.spongepowered.api.data.key.Key;
+import org.spongepowered.api.data.key.Keys;
+import org.spongepowered.api.data.manipulator.immutable.ImmutableDyeableData;
+import org.spongepowered.api.data.manipulator.mutable.DyeableData;
+import org.spongepowered.api.data.merge.MergeFunction;
+import org.spongepowered.api.data.type.DyeColor;
+import org.spongepowered.api.data.type.DyeColors;
+import org.spongepowered.api.data.value.BaseValue;
+import org.spongepowered.api.data.value.mutable.Value;
+import org.spongepowered.common.data.manipulator.immutable.ImmutableSpongeDyeableData;
+import org.spongepowered.common.data.manipulator.mutable.SpongeDyeableData;
+import org.spongepowered.common.data.processor.common.AbstractSingleDataProcessor;
+import org.spongepowered.common.data.value.immutable.ImmutableSpongeValue;
+import org.spongepowered.common.interfaces.block.tile.IMixinTileEntityShulkerBox;
+
+import java.util.Optional;
+
+public class ShulkerBoxDyeableDataProcessor extends AbstractSingleDataProcessor<DyeColor, Value<DyeColor>, DyeableData, ImmutableDyeableData> {
+
+    public ShulkerBoxDyeableDataProcessor() {
+        super(Keys.DYE_COLOR);
+    }
+
+    @Override
+    protected DyeableData createManipulator() {
+        return new SpongeDyeableData(DyeColors.PURPLE);
+    }
+
+    @Override
+    public boolean supports(DataHolder dataHolder) {
+        return dataHolder instanceof TileEntityShulkerBox;
+    }
+
+    @Override
+    public Optional<DyeableData> from(DataHolder dataHolder) {
+        if (supports(dataHolder)) {
+            return Optional.ofNullable(((IMixinTileEntityShulkerBox) dataHolder).getColor())
+                    .map(enumType -> (DyeColor) (Object) enumType)
+                    .map(SpongeDyeableData::new);
+        }
+        return Optional.empty();
+    }
+
+    @Override
+    public Optional<DyeableData> fill(DataContainer container, DyeableData dyeableData) {
+        Optional<String> id = container.getString(Keys.DYE_COLOR.getQuery());
+        if (id.isPresent()) {
+            dyeableData.set(Keys.DYE_COLOR, Sponge.getGame().getRegistry().getType(DyeColor.class, id.get()).get());
```

Comment:
Multiple points of failure in one line. Split this up into multiple lines to be able to throw exceptions as necessary.